{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Welcome to CASdesigner beta\n",
    "\n",
    "by Leo d'Espaux <leodespaux@gmail.com>\n",
    "help by William Zhuang\n",
    "\n",
    "Current as of 3/30/2016\n",
    "\n",
    "For a detailed user manual and feedback, please go to \n",
    "https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\n",
    "\n",
    "\n",
    "##Instructions\n",
    "1. Click on \"CLICK HERE\" in the cell below\n",
    "2. Click on the \"play\" button in the top bar\n",
    "\n",
    "    You may need to wait a minute. \n",
    "    \n",
    "    If you get stuck at any time, click the \"restart\" button, restart kernel, repeat 1-2.\n",
    "    \n",
    "    Then follow the simple in-line instructions.\n",
    "\n",
    "##Good luck!\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version 3.4.3 |Anaconda 2.3.0 (x86_64)| (default, Mar  6 2015, 12:07:41) \n",
      "[GCC 4.2.1 (Apple Inc. build 5577)]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Hi, welcome to CASdesigner.\n",
      "\n",
      "Refer to this community google doc for documentation and feedback:\n",
      "https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "WHAT DO YOU WANT TO DO TO S288C? Type \"1\", \"2\", or \"3\".\n",
      "\n",
      "\t 1 \t add to empty (insert DNA into a well-characterized site)\n",
      "\n",
      "\t 2 \t edit existing (delete or replace a gene of interest)\n",
      "\n",
      "\t 3 \t stitch (get primers to stich together N fragments)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "CLICK HERE \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "\n",
    "# import libraries we're using\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import copy\n",
    "from intermine.webservice import Service\n",
    "from pandas import *\n",
    "from pandas import DataFrame, read_csv\n",
    "import pandas as pd  \n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "from IPython.display import Image\n",
    "from IPython.display import display\n",
    "%matplotlib inline \n",
    "\n",
    "\n",
    "\n",
    "# define global variables\n",
    "HomologyLength = 1000\n",
    "PrimerMaxTm = 55 # As calculated by statluc, more like 65C by Thermo's Phusion program\n",
    "PrimerMaxLen = 60\n",
    "OverhangMaxFrac = 1\n",
    "\n",
    "\n",
    "\n",
    "def askUser():\n",
    "    print('Python version ' + sys.version+\"\\n\\n\\n\\n\")\n",
    "    print(\"\\nHi, welcome to CASdesigner.\\n\")\n",
    "\n",
    "          \n",
    "    print(\"Refer to this community google doc for documentation and feedback:\")\n",
    "    print(\"https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\")\n",
    "    print(\"\\n\\n\\n\\nWHAT DO YOU WANT TO DO TO S288C? Type \\\"1\\\", \\\"2\\\", or \\\"3\\\".\\n\")\n",
    "    print(\"\\t 1 \\t add to empty (insert DNA into a well-characterized site)\\n\") \n",
    "    print(\"\\t 2 \\t edit existing (delete or replace a gene of interest)\\n\")\n",
    "    print(\"\\t 3 \\t stitch (get primers to stich together N fragments)\\n\")\n",
    "\n",
    "   \n",
    "    Action= input(\"Your answer: \")\n",
    "    if Action == \"1\":\n",
    "        editEmpty()\n",
    "        \n",
    "    elif Action == \"2\":\n",
    "        editExisting()\n",
    "\n",
    "    elif Action == \"3\": # This is more for stitching things together not necessarily for CRISPR\n",
    "        buildCustom()\n",
    "             \n",
    "            \n",
    "            \n",
    "def editEmpty():    \n",
    "    \n",
    "    # I have an excel file where we keep cut site information\n",
    "    # and load it as a pandas dataframe called cutFrame\n",
    "    \n",
    "    cutFrame = pd.read_excel(\"locusTable.xlsx\", index_col=\"cutName\")\n",
    "    \n",
    "    #print(cutFrame)\n",
    "    \n",
    "    \n",
    "    LocusPic=Image(\"LocusLevelExp.png\")\n",
    "    display(LocusPic)\n",
    "    \n",
    "    \n",
    "    cutName=input(\"WHICH CUT SITE DO YOU WANT from the above list? e.g., \\\"208a\\\":\")\n",
    "\n",
    "    # cutName is a string that should correspond to one of the variables\n",
    "    # of the \"cutName\" column in dataFrame, e.g., 208a\n",
    "    \n",
    "    # cutFrame has the 20nt sequence of the gRNA saved, which we load\n",
    "    # It also has some more information we'll get to later\n",
    "    cutSequence=cutFrame.loc[cutName,'cutSequence']\n",
    "\n",
    "    # We need to load the chromosome where the cut site is\n",
    "    location=cutFrame.loc[cutName,'cutChr']+\".fasta\"\n",
    "     \n",
    "    ChromosomeSeq=SeqIO.read(location, \"fasta\").seq\n",
    "    \n",
    "    if ChromosomeSeq.find(cutSequence)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "        if ChromosomeSeq.find(cutSequence)==-1:\n",
    "            print(\"CAN'T FIND THE CUT SITE IN YOUR SEQUENCE\")\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(cutSequence)\n",
    "    EndIndex=StartIndex+34\n",
    "    \n",
    "    UpSeq=ChromosomeSeq[StartIndex-HomologyLength:StartIndex]\n",
    "    DownSeq=ChromosomeSeq[EndIndex:EndIndex+HomologyLength]\n",
    "        \n",
    "    UpHomRec = SeqRecord(UpSeq, name=\"up\"+cutName)\n",
    "    DownHomRec = SeqRecord(DownSeq, name=\"dn\"+cutName)\n",
    "    \n",
    "    print(\"\\n\\n\\n\\nWhat do you want to do to \" + cutName + \"?\\n\")\n",
    "    print(\"\\t1\\t You have a CDS but want help picking promoter and terminator\\n\")\n",
    "    print(\"\\t2\\t You already have a one-piece cassette and just want to integrate it\\n\\n\")\n",
    "    # Coming later\n",
    "    # print(\"\\t3\\t You want to construct a donor containing two standard cassettes (pointing away from each other\\n\")\n",
    "\n",
    "    typeEdit=input(\"Your answer: \")\n",
    "    if typeEdit==\"1\":\n",
    "    \n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        \n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec]\n",
    "\n",
    "    \n",
    "    elif typeEdit == \"2\":\n",
    "        \n",
    "        orfName=input(\"What's the name of your custom gene or cassette?\")\n",
    "        orfSeq=Seq(input(\"What's the sequence? \"))\n",
    "        orfRecord=SeqRecord(orfSeq,name=orfName)\n",
    "        \n",
    "        fragments=[UpHomRec, orfRecord, DownHomRec]\n",
    "\n",
    "    elif typeEdit == \"3\":\n",
    "        \n",
    "        print(\"Your construct will look like this: \")\n",
    "        print(\"up hom-<terminator1-gene1-promoter1<->promoter2-gene2-terminator2>-downhom\")\n",
    "        print(\"OK, let's build your FIRST cassette\")\n",
    "        PromoterRec1, orfRecord1, TerminatorRec1 = buildCassette()\n",
    "         \n",
    "        #now in the construction, these will be in the antisense direction, so let's change their sequences\n",
    "        rTerminatorRec1=flipRecord(TerminatorRec1)\n",
    "        rorfRecord1=flipRecord(orfRecord1)\n",
    "        rPromoterRec1=flipRecord(PromoterRec1)\n",
    "        \n",
    "        print(\"OK, let's build your SECOND cassette\")\n",
    "        PromoterRec2, orfRecord2, TerminatorRec2 = buildCassette()\n",
    "\n",
    "        fragments=[UpHomRec, rTerminatorRec1, rorfRecord1, rPromoterRec1,PromoterRec2,orfRecord2, TerminatorRec2,DownHomRec]\n",
    "\n",
    "    print(\"\\n\\n\\n\\n\\n\\n\\n\\n\")\n",
    "    stitch(fragments)\n",
    "    \n",
    "    \n",
    "    \n",
    "def editExisting():\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"WHICH GENE DO YOU WANT TO EDIT? Tell me a common name, e.g., \\\"OAF1\\\"\")\n",
    "    print(\"I'm smart and pretty and I can fetch it for you.\\n\")\n",
    "    \n",
    "    GeneName= input(\"Your answer: \")\n",
    "    \n",
    "    OrigGeneRecord=fetchGene(GeneName)\n",
    "    # note that this returns a seqrecord\n",
    "    \n",
    "    print(\"\\n\"+GeneName+\" is \"+OrigGeneRecord.features+\", which encodes \"+OrigGeneRecord.description+\", Bold choice.\")\n",
    "\n",
    "    \n",
    "    # We make seqrecords since that's what we carry through later in the program\n",
    "\n",
    "    UpHomRec = fetchNeighbor(OrigGeneRecord, \"upstream\", HomologyLength )\n",
    "    UpHomRec.name=\"up\"+UpHomRec.name \n",
    "    DownHomRec = fetchNeighbor(OrigGeneRecord, \"downstream\", HomologyLength )\n",
    "    DownHomRec.name=\"down\"+DownHomRec.name \n",
    "\n",
    "    print(\"\\n\\n\\n\\nWHAT DO YOU WANT TO DO TO \" + GeneName + \"?\\n\")\n",
    "    print(\"\\t1\\t DELETE the CDS cleanly--from start to stop codon\\n\")\n",
    "    print(\"\\t2\\t replace the CDS with another CDS\\n\")\n",
    "    print(\"\\t3\\t replace the CDS with an expression cassette I will help you build\\n\")\n",
    "    print(\"\\t4\\t replace a specified region near your target gene with another specified region\\n\")\n",
    "    \n",
    "    typeEdit=input(\"Your answer: \")\n",
    "    \n",
    "    # Now let's pick the cut site. This is independent from the donor cassette\n",
    "    # We have an excel sheet with genes and pre-picked cut sites\n",
    "    # Or let the user pick\n",
    "    \n",
    "    print(\"\\n\\n\\n\\nHave you heard of CRISPR? We will need a guideRNA. In the future I will help you pick it.\")\n",
    "    cutChoice=input(\"Please enter your 20-nt gRNA here.\\n\")\n",
    "    \n",
    "    \n",
    "    if cutChoice==\"p\":\n",
    "        cutSequence=cutFinder(GeneName)\n",
    "    elif len(cutChoice)==20:\n",
    "        cutSequence=cutChoice\n",
    "        # Later should check that the sequence is present and unique\n",
    "    \n",
    "   \n",
    "    \n",
    "    #note that in all the below, we want to have fragments be records\n",
    "    if typeEdit==\"1\":\n",
    "        fragments=[UpHomRec,DownHomRec]\n",
    "    \n",
    "    if typeEdit==\"3\":\n",
    "        PromoterRec, orfRecord, TerminatorRec = buildCassette()\n",
    "        fragments=[UpHomRec, PromoterRec, orfRecord, TerminatorRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    if typeEdit==\"2\":\n",
    "        print(\" \")\n",
    "        NewGeneName=input(\"\\n\\n\\n\\nWhat's the name of the gene you're inserting?\")\n",
    "        NewGeneSeq=Seq(input(\"\\n\\n\\n\\nWhat's the sequence of your new gene? \"))\n",
    "        InsertRec = SeqRecord(NewGeneSeq, name=NewGeneName)\n",
    "        fragments=[UpHomRec, InsertRec, DownHomRec]\n",
    "\n",
    "    if typeEdit==\"4\":\n",
    "        print(\"How many pieces (other than homology fragments) are you stitching together.\")\n",
    "        Npieces=input(\"Your answer: \")\n",
    "        output = buildCassette()\n",
    "\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    \n",
    "    print(\"\\n\\n\\n\\n\\n\\n\\n\\nHere is the oligo to clone or gap-repair your cut plasmid:\\n\")\n",
    "    print(\"cut\"+GeneName+\" tggcgaatgggacttt\"+cutSequence+\"gttttagagctagaaatagcaagt\\n\")\n",
    "\n",
    "    # Now we feed the design to a tool to pick primers which amplify each fragment\n",
    "    # and add terminal homology to the neighboring fragments\n",
    "    \n",
    "    \n",
    "    stitch(fragments)\n",
    "\n",
    "    \n",
    "    \n",
    "def buildCustom():\n",
    "    N = int(input(\"How many pieces in your custom cassette: \"))\n",
    "    fragments = variableCassette(N)[0]\n",
    "    stitch(fragments)\n",
    "\n",
    "            \n",
    "    #------------------------------ FETCH FUNCTIONS -------------------------------------\n",
    "\n",
    "    \n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "       \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "    GeneRecord.description=descr\n",
    "\n",
    "    return GeneRecord \n",
    "       \n",
    "    \n",
    "    \n",
    "def fetchNeighbor(NeighborRecord, direction, distance):\n",
    "\n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if NeighborRecord.features[1]==\"A\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"B\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer02.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"C\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer03.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"D\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer04.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"E\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer05.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"F\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer06.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"G\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer07.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"H\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer08.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"I\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer09.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"J\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer10.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"K\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer11.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"L\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer12.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"M\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer13.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"N\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer14.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"O\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer15.fasta\", \"fasta\")\n",
    "    if NeighborRecord.features[1]==\"P\":\n",
    "        ChromosomeRec=SeqIO.read(\"Scer16.fasta\", \"fasta\") \n",
    "\n",
    "    \n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    NeighborSeq=NeighborRecord.seq\n",
    "    ChromosomeSeq=ChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if ChromosomeSeq.find(NeighborSeq)==-1:\n",
    "        ChromosomeSeq=ChromosomeSeq.reverse_complement()\n",
    "\n",
    "    StartIndex=ChromosomeSeq.find(NeighborSeq)\n",
    "    EndIndex=StartIndex+len(NeighborSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        DesiredSeq=ChromosomeSeq[StartIndex-distance:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        DesiredSeq=ChromosomeSeq[EndIndex:EndIndex+distance]\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    NeighborRec = SeqRecord(DesiredSeq, name=NeighborRecord.name)\n",
    "    \n",
    "    return NeighborRec\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    #------------------------------------ CONSTRUCTING STUFF --------------------------------------\n",
    "\n",
    "def getPrimer(currRecord):\n",
    "    \n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    seq=currRecord.seq\n",
    "    \n",
    "    while mp <= PrimerMaxTm and length <= PrimerMaxLen:\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_staluc(primer)\n",
    "        length += 1\n",
    "\n",
    "    return primer           \n",
    "        \n",
    "def overhangPrimer(currRecord,prevSeq):\n",
    "    #let's get the template-binding primer first\n",
    "    primer=getPrimer(currRecord)\n",
    "    \n",
    "    \n",
    "    #OK let's work on the overhang\n",
    "    maxOhLen=PrimerMaxLen-len(primer)    \n",
    "    maxFrac=1\n",
    "    \n",
    "    #let's decide on a max overhang length\n",
    "    if round(len(primer)*(OverhangMaxFrac+1)) < 60:\n",
    "             maxOhLen=round(len(primer)*OverhangMaxFrac)\n",
    "    \n",
    "    #the index must be an integer!!!\n",
    "    maxOhLen=int(maxOhLen)\n",
    "    ohprimer=prevSeq.seq[-maxOhLen:]+primer #we add the .seq so that it returns a string\n",
    "    \n",
    "    return ohprimer      \n",
    "        \n",
    "\n",
    "          \n",
    "def buildCassette():\n",
    "    \n",
    "    # Let's display data\n",
    "    PromPic=Image(\"PromoterLevelExp.png\")\n",
    "    display(PromPic) \n",
    "       \n",
    "    PromoterName=input(\"Which PROMOTER do you want to use, e.g., \\\"TDH3\\\": \")\n",
    "    PromoterGeneRec=fetchGene(PromoterName)\n",
    "    PromoterRec=fetchNeighbor(PromoterGeneRec,\"upstream\",600)\n",
    "    PromoterRec.name=PromoterRec.name+\"ps\"\n",
    "    \n",
    "    \n",
    "    #second, the terminator\n",
    "    TerminatorName = input(\"\\n\\n\\n\\nWhich TERMINATOR do you want to use, e.g., \\\"ADH1\\\": \")\n",
    "    TerminatorGeneRec=fetchGene(TerminatorName)\n",
    "    TerminatorRec=fetchNeighbor(TerminatorGeneRec,\"downstream\",250)\n",
    "    TerminatorRec.name=TerminatorRec.name+\"ts\"\n",
    "    \n",
    "    #and last, the gene\n",
    "    orfName = input(\"\\n\\n\\n\\nWhat is the name of your CDS, e.g., \\\"KlGapDH\\\": \")\n",
    "    \n",
    "    orfSeq=input(\"\\n\\n\\n\\nWhat's the sequence: \")\n",
    "    \n",
    "    \n",
    "    print(\"\\n\\n\\n\\nWe have a couple of protein tags we can show you how to add.\")\n",
    "    print(\"Ask one of us if you're interested in this type of stuff.\")\n",
    "    LocTags=Image(\"LocTags.png\")\n",
    "    display(LocTags)\n",
    "    \n",
    "    \n",
    "    orfRecord=SeqRecord(Seq(orfSeq), name=orfName)\n",
    "    \n",
    "    #print(\"\\nDo you want to attach any protein tags?\\n\")\n",
    "    #print(\"This is under construction but here is tag CNE1 targeting to the ER lumen.\\n\")\n",
    "    #CNE1tag=Image(\"CNE1tag.png\")\n",
    "    #display(CNE1tag)\n",
    "    \n",
    "    insertRec=[PromoterRec,orfRecord,TerminatorRec]\n",
    "    return PromoterRec, orfRecord, TerminatorRec\n",
    "    \n",
    "\n",
    "    \n",
    "\n",
    "def variableCassette(N, toVary = 0, variants = 0):\n",
    "    print(\"\")\n",
    "    print(\"Let's start building.\")\n",
    "    print(\"\")\n",
    "    \n",
    "    # Store both name and sequence in a SeqRecord\n",
    "    # Append them to a list\n",
    "    # Return list as fragments to be stitched\n",
    "\n",
    "    records = []\n",
    "    for n in range(N):\n",
    "        name = input(\"What is the name of sequence \" + str(n+1) +\":\")\n",
    "        sequence = input(\"What is the sequence of this fragment:\\n\")\n",
    "        Rec = SeqRecord(Seq(sequence), id = str(n+1))\n",
    "        Rec.name = name\n",
    "        records.append(Rec)\n",
    "\n",
    "    variantRecords = []\n",
    "    variantRecords.append(records)\n",
    "    # This only happens if there are variants.\n",
    "    if variants > 0:\n",
    "        print(\"Time to make those variants you wanted.\")\n",
    "        for n in range(variants-1):\n",
    "            name = input(\"What is the name of variant \" + str(n+1) + \":\")\n",
    "            sequence = input(\"What is the sequence of this variant:\")\n",
    "            Rec = SeqRecord(Seq(sequence), id = str(n+1))\n",
    "            Rec.name = name\n",
    "            # Make a copy of the original, switch the fragments and add it to the list. \n",
    "            # Deep-copy ensures there are no pointer issues\n",
    "            tempVariant = copy.deepcopy(records)\n",
    "            tempVariant[toVary - 1] = Rec\n",
    "            variantRecords.append(copy.deepcopy(tempVariant))\n",
    " \n",
    "    # Returns a list of lists of the SeqRecords of the fragments\n",
    "    return variantRecords\n",
    "\n",
    "\n",
    "def stitch(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    \n",
    "    #let's make an empty sequence file\n",
    "    Nfrags=len(fragments)\n",
    "    donor=Seq(\"\")\n",
    "    index=[]\n",
    "\n",
    "    print(\"\\n\\n\\n\\nHere are the primers to amplify your fragments and construct your donor DNA cassette:\\n\")\n",
    "\n",
    "    for i in range (0, Nfrags):\n",
    "        donor=donor+fragments[i]\n",
    "    \n",
    "    # The names include information on the homology provided by the overhang\n",
    "    # Note that some primers don't have overhangs\n",
    "    for i in range (0, Nfrags):\n",
    "        if i==0:\n",
    "            print(\"F\"+ fragments[i].name + \" \" + getPrimer(donor))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "        elif i==Nfrags-1:\n",
    "            print(\"F\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \" \" + getPrimer(donor.reverse_complement()))\n",
    "        else:\n",
    "            print(\"F\"+ fragments[i].name + \"(\" + fragments[i-1].name + \") \" + overhangPrimer(fragments[i],fragments[i-1]))\n",
    "            print(\"R\"+ fragments[i].name + \"(\" + fragments[i+1].name + \") \" + overhangPrimer(fragments[i].reverse_complement(),fragments[i+1].reverse_complement()))\n",
    "\n",
    "    print(\"\\n\\nThe size and sequence of your donor DNA is below.\")\n",
    "\n",
    "    print(len(donor.seq))\n",
    "\n",
    "    print(donor.seq)\n",
    "\n",
    "    print(\"\\n\\n\\n\\nGOOD LUCK! Refer to the this link for user guide, feedback, and community tips.\")\n",
    "    print(\"https://docs.google.com/document/d/11YNoUBtit4rkWZQr_ZclzAOTE_HGMtgB3H_6cvq1Y24/edit?usp=sharing\\n\")\n",
    "    \n",
    "        \n",
    "def flipRecord(origRecord):\n",
    "    origRecord.seq=origRecord.seq.reverse_complement()\n",
    "    origRecord.name=\"r\"+origRecord.name\n",
    "    return origRecord\n",
    "    \n",
    "def cutFinder(GeneName):\n",
    "    cutFrame = pd.read_excel(\"cutsTable.xlsx\", index_col=\"cutName\")\n",
    "    \n",
    "    # cutFrame has the 20nt sequence of the gRNA saved, which we load\n",
    "    # It also has some more information we'll get to later\n",
    "    \n",
    "    cutSequence=cutFrame.loc[GeneName,'cutSequence']\n",
    "    \n",
    "    print(\"\\nI picked the following cut site for you \"+cutSequence)\n",
    "    \n",
    "    return cutSequence\n",
    "\n",
    "\n",
    "askUser()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "Hold on a moment, it's working... \n",
    "The interface should appear below momentarily.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
